### PR TITLE
[inductor] addmm + ReLU / GELU fusion pass

### PR DESCRIPTION
**Pull request by aakhundov**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104132

Summary:

Add a new path in `post_grad.py` for replacing addmm + ReLU / GELU activation with the corresponding `_addmm_activation` call (with `use_gelu=False` or `True`, respectively). The replacement is done only on `max_autotune_gemm=False` and when the activation is fusible.

Test Plan:

$ python test/inductor/test_pattern_matcher.py -k test_addmm_activation -v

(__main__.TestPaternMatcher.test_addmm_activation) ... /data/users/aakhundov/pytorch/torch/_inductor/compile_fx.py:128: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
  warnings.warn(
Using FallbackKernel: aten._addmm_activation.default
Using FallbackKernel: aten._addmm_activation.default
/data/users/aakhundov/pytorch/torch/_dynamo/eval_frame.py:373: UserWarning: changing options to `torch.compile()` may require calling `torch._dynamo.reset()` to take effect
  warnings.warn(
frames [('total', 1), ('ok', 1)]
stats [('calls_captured', 2), ('unique_graphs', 1)]
aot_autograd [('total', 1), ('ok', 1)]
inductor []
ok

----------------------------------------------------------------------
Ran 1 test in 13.415s

OK

Reviewers: @eellison

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8

### CODE REVIEW

**Comment by aakhundov** [comment]
The PR partially addresses #103480.

**Comment by aakhundov** [comment]
In this PR addressing #103480, I'm trying to register a new replacement pass, as a part of `post_grad` passes, to replace `addmm` + ReLU / GELU with `_addmm_activation`. The pass seems to work fine with static shapes, but when I set `dynamic=True` in `torch.compile`, I'm getting this error (similar to the one reported in #103194):

```
...
  File "/fsx/users/aakhundov/local/pytorch/torch/_inductor/compile_fx.py", line 268, in compile_fx_inner
    compiled_graph: CompiledFxGraph = fx_codegen_and_compile(
                                      ^^^^^^^^^^^^^^^^^^^^^^^
  File "/fsx/users/aakhundov/local/pytorch/torch/_inductor/compile_fx.py", line 450, in fx_codegen_and_compile
    post_grad_passes(gm, locality_reorder=locality_reorder)
  File "/fsx/users/aakhundov/local/pytorch/torch/_inductor/fx_passes/post_grad.py", line 64, in post_grad_passes
    patterns.apply(gm.graph)
  File "/fsx/users/aakhundov/local/pytorch/torch/_inductor/pattern_matcher.py", line 870, in apply
    if m and entry.extra_check(m):
             ^^^^^^^^^^^^^^^^^^^^
  File "/fsx/users/aakhundov/local/pytorch/torch/_inductor/pattern_matcher.py", line 695, in check_fn
    args[i] = torch.empty_strided(
              ^^^^^^^^^^^^^^^^^^^^
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
RuntimeError: /fsx/users/aakhundov/local/pytorch/build/aten/src/ATen/RegisterCUDA.cpp:7012: SymIntArrayRef expected to contain only concrete integers
```

The error happens at [this call](https://github.com/pytorch/pytorch/blob/b2277075b0fe5cf085d369a313863f64c6fb50c3/torch/_inductor/pattern_matcher.py#L695-L701) to `torch.empty_strided` in the `check_fn` intended to retrace the graph with the real input shapes (not the dummy ones used during the pattern registration). These are the values of [`args`](https://github.com/pytorch/pytorch/blob/b2277075b0fe5cf085d369a313863f64c6fb50c3/torch/_inductor/pattern_matcher.py#L685-L689) before the for-loop [here](https://github.com/pytorch/pytorch/blob/b2277075b0fe5cf085d369a313863f64c6fb50c3/torch/_inductor/pattern_matcher.py#L690):

```
[FakeTensor(..., device='cuda:0', size=(s2,)),
 FakeTensor(..., device='cuda:0', size=(s0, s1)),
 FakeTensor(..., device='cuda:0', size=(s1, s2))]
```

So it seems that `torch.empty_strided` is not happy with the symbolic shape passed in `sizes`.

Interestingly, this exact call to `torch.empty_strided` runs successfully on symbolic shapes in the `check_fn` run for the `pad_mm` replacement passes [registered](https://github.com/pytorch/pytorch/blob/b2277075b0fe5cf085d369a313863f64c6fb50c3/torch/_inductor/fx_passes/pad_mm.py#L395) as a part of `joint_graph`. Does this mean there is some difference in the global context that changes the behavior of `torch.empty_strided` on symbolic shapes between the `joint_graph` and `post_grad` passes?

Minimal repro on top of this PR:

```
import torch

def fn(mat1, mat2, inp):
    linear = torch.addmm(inp, mat1, mat2)
    result = torch.nn.functional.gelu(linear)
    return (result,)

mat1 = torch.randn([3, 4]).cuda()
mat2 = torch.randn([4, 5]).cuda()
inp = torch.randn([5]).cuda()
y = fn(mat1, mat2, inp)[0]

fn_inductor = torch.compile(dynamic=True)(fn)
y_inductor = fn_inductor(mat1, mat2, inp)[0]
```

cc @eellison @ezyang. Thanks a lot!

File: torch/_inductor/pattern_matcher.py [hunk 1]
```diff
@@ -692,13 +693,14 @@ def check_fn(match: Match):
...
                     return False
 
-                args[i] = torch.empty_strided(
-                    args[i].size(),
-                    args[i].stride(),
-                    dtype=args[i].dtype,
-                    device=args[i].device,
-                    requires_grad=grad,
-                )
+                with torch._dynamo.utils.detect_fake_mode(args):
```
Comment (ipiszy) [inline torch/_inductor/pattern_matcher.py:696]: Wonder what does this mean?

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -1189,6 +1189,19 @@ def addmm(self: Tensor, mat1: Tensor, mat2: Tensor, beta: int = 1, alpha: int =
     return out + beta * self
 
 
+@register_decomposition(aten._addmm_activation)
```
Comment (ipiszy) [inline torch/_decomp/decompositions.py:1192]: Wonder why do we need this decomposition pass? It seems to me that when max_auto_tune is false, we do fusion via the pattern-matching fx pass. When max_auto_tune is True, fusion is not done and we also don't need this decomposition pass, correct?

**Review by ipiszy — COMMENTED** [review]
Thanks @aakhundov @eellison! I have some dumb questions, see inline.

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (eellison) [reply torch/_decomp/decompositions.py:1192]: This is registering the decomposition so that we get meta support with symbolic sizes, but it's not enabling it for torch/_inductor. To be enabled it would need to be in the list [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/decomposition.py#L40) or in core_aten_decomps.

We reuse the decompositions for fake tensor meta kernel propagation. Since there's already an addmm aten decomp, adding addmm_activation decomp is easiest way to get meta kernel support.

Comment (ipiszy) [reply torch/_decomp/decompositions.py:1192]: I see, thanks @eellison !

File: torch/_inductor/pattern_matcher.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/pattern_matcher.py:696]: it's the api to grab FakeTensorMode either from the inputs, from the currently enabled FakeTensorMode on the stack, or from the current TracingContext, which includes things like guards/etc that is generated from dynamo. It makes it so that the empty_strided tensor will be a fake tensor, so that it doesnt error when trying to instantiate a tensor with symbolic shapes as size. 

FakeTensors correspond to a single fake tensor mode. It is not intended to use FakeTensors from different modes, because the Mode has things like cached storage conversions for correctness. So we reuse the existing mode instead of creating a new one.

File: test/inductor/test_pattern_matcher.py [hunk 3]
```diff
@@ -84,6 +84,58 @@ def fn(a, b, c):
...
+            (fn_addmm_relu, 2, 1e-8),
+            # higher tolerance due to the "tanh" approximation
+            # in fused GELU epilogue vs. "none" without fusion
+            (fn_addmm_gelu, 6, 1e-3),
+        ):
+            counters.clear()
+            expected = fn(*args)
+            actual = torch.compile(fn)(*args)
+            torch.testing.assert_close(actual, expected, atol=atol, rtol=0)
+            self.assertEqual(counters["inductor"]["pattern_matcher_count"], 1)
```
Comment (eellison) [inline test/inductor/test_pattern_matcher.py:110]: Can also test that it is hit with the `run_and_get_code` api, and then grep for "addmm_activation"

**Review by eellison — APPROVED** [review]
Sweet! Looks great! We should check that we don't regress regress inference accuracy with this on for the gelu appoximation. I think you have a benchmarking run on https://hud.pytorch.org/benchmark/compilers. If that is the case we can selectively apply the gelu transformation if user enables it, or only match against the gelu with "tanh" approximation passed in. 

We can make this more visible in the freezing api when that is officially released.

**Comment by aakhundov** [comment]
@eellison @Chillee Seems that the current state of the PR passes the CI (+ trunk / periodic).

I've also run the benchmarking job: in the https://hud.pytorch.org/benchmark/compilers, when I select the "inference" / "bfloat16", then my branch (`gh/aakhundov/6/head`) on the left and a commit from `main` shortly before or after on the right, I don't see any substantial differences on the front page. Does this mean that the inference accuracy is not regressed? Or are there more specific places to check that in the dashboard?

In other words, from the test and benchmarking results, I'm trying to understand if this PR is good to merge as is or if we need to add a gating config for the transformation and / or `approximate="tanh"` to the pattern. Thanks!

**Comment by aakhundov** [comment]
@jansel does the PR look good to merge? Or do you think we should put the new pass behind a config flag (disabled by default) / add `approximate="tanh"` to the pattern? Thanks!

**Comment by eellison** [comment]
For further context - all of our accuracy tests passed, and we'd only be enabling this for inference since there are other fusions possible in training. it does use a tanh gelu approximation. We got a 1-2% HF inference win with this pr.

**Comment by Chillee** [comment]
IMO, we can merge it as is, but I think it'd be good to have a follow up that we only match if `approximate="tanh"`. We could also maybe throw a warning if we would match the pattern except for the fact that their gelu doesn't support approximation.

One thing I'm confused about though - shouldn't we already be decomposing gelu as part of the decompositions? How does Inductor see this at the pattern matching phase if it's post-grad?

**Comment by aakhundov** [comment]
@Chillee Re. your last question, my understanding is that the pattern is also decomposed at some point before being matched. So when the pattern is applied after graph decompositions (e.g., as a part of the joint graph or post-grad passes), it matches the decomposed GELU successfully. @eellison can share more specific details.

**Comment by Chillee** [comment]
@aakhundov Oh right, I didn't realize that that feature of our pattern matcher could be used like this 🤔

File: torch/_inductor/fx_passes/post_grad.py [hunk 4]
```diff
@@ -344,6 +353,76 @@ def addmm(match, mat1, mat2, inp):
...
+            replacement,
+            args,
+            inference_graph,
+            inference_patterns,
+            extra_check=should_replace_addmm_activation,
+        )
+
+    counters[
+        "inductor"
+    ] = counters_ref  # clear view matches encountered during mm tracing
```
Comment (jansel) [inline torch/_inductor/fx_passes/post_grad.py:423]: Can we move this counter checkpointing to the `@init_once_fakemode` decorator?

**Review by jansel — APPROVED** [review]
